### ISSUE TITLE
Question: How to enable dual-stack for the listener address?

### ISSUE THREAD

**Issue by lolin9** [envoyproxy/envoy#11124]
When I apply `0.0.0.0` to  the listener's address`SocketAddress` it works only on IPv4 and if `::` is used then only IPv6 works. How to enable both in the same listener?

**Comment by mattklein123** [comment]
Unfortunately the only way to do this today is to duplicate the entire listener. This is something that I would like to consider changing (allowing a listener to listen on multiple sockets) as this has come up a few times. cc @lambdai for any thoughts on this.

**Comment by lolin9** [comment]
Thanks for the response @mattklein123. I think you could introduce a new field `addresses` in `Listener`. This field is an array of `core.v3.Address` and can override the `address` field only if it is populated. Of course this is just a quick fix at the API level, but I am not sure whether the internals of the listener implementation can make it possible or this will require a huge amount of work.

**Comment by lambdai** [comment]
I haven't dig into it. I thought it was libevent cannot bind to both v6 and v4 at the same time. I was wrong: by default :: is bind to both 0.0.0.0 and :: in libevent

I not sure what is blocking envoy from bind to both

**Comment by lambdai** [comment]
I tried to understand where is the source of truth of that IPV6_V6ONLY sockopt is injected and i realize we could clear it...
```
42547 setsockopt(15, SOL_IPV6, IPV6_V6ONLY, [0], 4) = 0

```
```
$ ss -ntlp |grep 9998
LISTEN   0         128                    [::]:9998                [::]:*        users:(("envoy-static",pid=42305,fd=15)) 
$ ss -nltp |grep 9998
LISTEN   0         128                       *:9998                   *:*        users:(("envoy-static",pid=42547,fd=15))
```

Try below snippet of listener config
```
$ head -10 *v4v6*
static_resources:
  listeners:
  - address:
      socket_address:
        address: "::"
        port_value: 9998
        ipv4_compat: true
```

**Comment by lambdai** [comment]
tl;dr
set ipv4_compat in listener address
https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/core/address.proto#core-socketaddress

**Comment by mattklein123** [comment]
Oh nice find, thanks @lambdai. @lolin9 is it still worth tracking allowing a listener to bind to multiple addresses/sockets or is this good enough for your use?

**Comment by lolin9** [comment]
@mattklein123 and @lambdai thank you both, that certainly solves it. I actually tried it before opening this issue but when I used `netstat` to check whether the ipv4 socket is open I saw nothing. I hope you still consider having multiple addresses per listener since the `0.0.0.0` is a special case here. Duplicating the entire listener especially when having complex configuration with lots of HTTP filters, routes, RBACs,... will eventually affect the performance and memory usage.
